lib/elixir/lib/kernel.ex

```diff
@@ -4956,6 +4955,10 @@ defmodule Kernel do
   defp expand_module_alias(other, env), do: Macro.expand(other, env)
 
   # defmodule Elixir.Alias
+  defp alias_defmodule({:__aliases__, _, [:"Elixir", _]}, _module, _env),
+    do: nil
+
+  # defmodule Elixir.Alias.Nested...
```

Comment:
```
  defp alias_defmodule({:__aliases__, _, [h | t]}, module, %{module: env_module})
       when is_atom(h) and h != :"Elixir" and is_atom(module) and env_module != nil do
    module = :elixir_aliases.concat([env_module, h])
    alias = String.to_atom("Elixir." <> Atom.to_string(h))
    case t do
      [] -> {module, module, alias}
      _ -> {String.to_atom(Enum.join([module | t], ".")), module, alias}
    end
  end

  defp alias_defmodule(_ast, _module, _env) do
    false
  end
```

This way you can also get rid of the `with`.

Reply:
Oh you're right in the case of nesting, I misunderstood.

Ignoring in all cases including the `defp alias_defmodule(_raw, module, _env) do` was my original approach, but it was emitting warnings so I went for something more precise:

<img width="944" alt="Screenshot 2023-10-17 at 7 42 59" src="https://github.com/elixir-lang/elixir/assets/11598866/bdcf0927-c1c7-43cf-a438-b49e1ce69692">

I should be able to replace the `with` by a simple `case` though.

Reply:
https://github.com/elixir-lang/elixir/blob/main/lib/elixir/test/elixir/module_test.exs#L317-L334

`raw` is `ModuleTest.RawModule` or `{:__aliases__, [line: 328, column: 25],
 [{:__MODULE__, [line: 328, column: 15], nil}, :NonAtomAlias]}`

Reply:
How about [9c33fef](https://github.com/elixir-lang/elixir/pull/13011/commits/9c33fef7667b2d6ee021de273793cf01d73cb901)?

Reply:
LGTM, I would just return `false` instead of `nil` but it is minor. :)

Reply:
Actually, nevermind, the current fix is wrong. I can add a test similar to the other warnings that show it:

```elixir
  test "does not leak alias from Elixir root alias" do
    defmodule Elixir.ModuleTest.ElixirRootAlias do
      def hello, do: :world
    end

    refute __ENV__.aliases[Elixir.ModuleTest]
    refute __ENV__.aliases[Elixir.ElixirRootAlias]
    assert Elixir.ModuleTest.ElixirRootAlias.hello() == :world
  end
```

Reply:
And I think the proper fix is this one:

```
  defp alias_defmodule({:__aliases__, _, [h | t]}, module, %{module: env_module})
       when is_atom(h) and h != :"Elixir" and is_atom(module) and env_module != nil do
    module = :elixir_aliases.concat([env_module, h])
    alias = String.to_atom("Elixir." <> Atom.to_string(h))

    case t do
      [] -> {module, module, alias}
      _ -> {String.to_atom(Enum.join([module | t], ".")), module, alias}
    end
  end

  defp alias_defmodule(_ast, module, _env) do
    {module, module, nil}
  end
```

That's because we are using the alias to define a context module which is used by the compiler to avoid warnings. Perhaps we should stop using the alias like this, but it is how it is right now.

Reply:
> I would just return false instead of nil but it is minor. :)

Just for my reference, what would have been the rationale to pick `false` in this case?
I don't recall seeing `tuple | false` very frequently except some erlang bifs, so I simply wondered what's the rule of thumb to pick `false` over `nil`.

Reply:
I find `false` more intentional. There are some constructs that return `nil` by default (for example `if` without `else`), so you may have unexpected `nil`s.

Reply:
Thanks for clarifying!